src/mrpro/algorithms/_cg.py

```diff
@@ -0,0 +1,124 @@
+    max_iterations: int = 128,
+    tolerance: float = 1e-4,
+    callback: Callable | None = None,
+) -> torch.Tensor:
+    """CG for solving a linear system Hx=b.
+
+    Thereby, H is a linear self-adjoint operator, b is the right-hand-side
+    of the system and x is the sought solution.
+
+    Note that this implementation allows for simultaneously solving a batch of B problems
+    of the form
+        H_i x_i = b_i,      i=1,...,B.
+    Thereby, the underlying assumption is that the considered problem is H x = b with
+        H:= diag(H_1, ..., H_B), b:= [b_1, ..., b_B]^T.
+    Thus, if all H_i are self-adjoint, so is H and the CG can be applied.
+    Note however, that the accuracy of the obtained solutions might vary among the different
+    problems.
+
+    Also, note that if the condition of H is very large, a small residual does not necessarily
+    imply that the solution is accurate.
+
+    Parameters
+    ----------
+    operator
+        self-adjoint operator (named H above)
+    right_hand_side
+        right-hand-side of the system (named b above)
+    starting_value, optional
+        starting value of the iteration; if None, it will be set to b
+    max_iterations, optional
+        maximal number of iterations
+    tolerance, optional
+        tolerance for the residual
+    callback, optional
+        user-provided function to be called at each iteration
+
+    Returns
+    -------
+        an approximate solution of the linear system Hx=b
+    """
+
+    # throw an error if the shapes of the starting value and the right-hand side
+    # do not coincide
+    if starting_value is not None:
+        if starting_value.shape != right_hand_side.shape:
+            raise ValueError(f'\nshapes {starting_value.shape, right_hand_side.shape} are incompatible')
+
+    # initial residual
+    residual = right_hand_side - operator(starting_value)[0] if starting_value is not None else right_hand_side.clone()
+
+    # initialize conjugate vector
+    conjugate_vector = residual.clone()
+
+    # assign starting value to the solution
+    solution = starting_value.clone() if starting_value is not None else right_hand_side.clone()
+
+    # for the case where the residual is exactly zero
+    if torch.vdot(residual.flatten(), residual.flatten()) == 0:
+        return solution
+
+    # squared tolerance;
+    # (we will check ||residual||^2 < tolerance^2 instead of ||residual|| < tol
+    # #to avoid the computation of the root for the norm)
+    tolerance_squared = tolerance**2
+
+    # dummy value the old squared norm of the residual;
+    # #only required for initialization
+    norm_squared_residual = None
+
+    for i in range(max_iterations):
+
+        # calculate the square norm of the residual
+        residual_flat = residual.flatten()
+        norm_squared_residual_new = torch.vdot(residual_flat, residual_flat).real
+
+        # check if the solution is already accurate enough
+        if tolerance != 0:
```

Comment:
Does this mean that for tolerance = 0 the algorithm will always run for max_iterations even if norm_squared_residual_new == 0? Is this what we want?

Reply:
this is related to @fzimmermann89's comment above where he mentioned that cuda will synchronize after each check. 

i think it makes sense to keep this logic. The tolerance and the maximal number of iterations are two stopping criterions and if either of the two is met, the iteration stops.

Setting tolerance=0 therefore corresponds to only having the maximal number of iterations as a stopping criterion. 
I would say this is the desired behaviour of the algorithm regardless of whether or not this corresponds to what one would do in practice.

having a exactly zero residual is very unlikely, I would say. Even in the absence of noise, rounding errors will proably avoid having ||r||==0. The check in the beginning before the iterations is to prevent the iterations to start if we start with the exact solution (also something that would probably not occur in practice).

Reply:
tolarance == 0 meaning: dont stop on a tolerance criterion feels quite logical to me.
maybe we can also explicitly mention this in the docstring

Reply:
> maybe we can also explicitly mention this in the docstring

yes, please